File: ICPC2019/BUAA Summer Training 6/d.cpp
#include <bits/stdc++.h>
#define ll long long
#define N 110
using namespace std;

ll a[N],b[N];
ll m,k,r;
int n;

ll exgcd(ll a,ll b,ll &x,ll &y)
{
	if (!b) {x=1; y=0; return a;}
	ll d=exgcd(b,a%b,x,y); 
	ll t=x; x=y; y=t-a/b*y;
	return d;
}

int check(int i,int j)
{
	ll x,y;
	ll d=exgcd(a[i],a[j],x,y);
	if ((b[i]-b[j])%d) return 1;
}

int main()
{
	freopen("1.in","r",stdin);
	scanf("%d %lld",&n,&m);
	for (int i=1;i<=n;i++) scanf("%d %d",&a[i],&b[i]);
	for (int i=1;i<=n;i++) for (int j=i+1;j<=n;j++) if (check(i,j))
	{
		puts("he was definitely lying");
		return 0;
	}
	k=a[1]; r=b[1];
	for (int i=2;i<=n;i++)
	{
		ll x,y; ll d=exgcd(k,a[i],x,y);
		x=(x*(r-b[i])/d+a[i]/d)%(a[i]/d); y=k/d*a;
		r=(x*k+r)%y; 
		k=y;
	}
}

File: ICPC2019/BUAA Summer Training 6/d.py
def exgcd(a, b):
    if b==0:
        return 1, 0, a
    x, y, d = exgcd(b, a%b)
    t = x
    x = y
    y = t-a//b*y
    return x, y, d 

def check(i, j):
    x, y, d = exgcd(a[i], a[j])
    if (b[i]-b[j])%d != 0:
        return 1
    return 0

n, m=input().split()
n = int(n)
m = int(m)
# print(n**100)
a, b=[0], [0]
for i in range(1, n+1):
    x, y=input().split()
    x = int(x)
    y = int(y)
    a.append(x)
    b.append(y)
# print(a)
for i in range(1, n+1):
    for j in range(i+1,n+1):
        if check(i,j):
            print("he was definitely lying")
            exit (0)
k=a[1]
r=b[1]
for i in range(2, n+1):
    x, y, d = exgcd(k, a[i])
    x=(x*(b[i]-r)//d+a[i]//d)%(a[i]//d)
    y=k//d*a[i]
    # print(k, r, x, y)
    r=(x*k+r)%y
    k=y
if r>m :
    print("he was probably lying")
else:
    print(r)
        
